File: src/lib/tauri/cloud-policy.ts
import { invoke } from "@tauri-apps/api/core";

export type CloudPolicyBindingRequest = {
  organizationId: string;
  profileIds: string[];
  supabaseUrl: string;
  publishableKey: string;
  accessToken: string;
  expiresAt: number;
};

export type CloudPolicyCredentialRequest = Omit<CloudPolicyBindingRequest, "organizationId" | "profileIds">;
export type CloudPolicyStatus = {
  organizationId: string;
  enabled: boolean;
  authenticated: boolean;
  profileCount: number;
};

export const bindCloudPolicy = (request: CloudPolicyBindingRequest) =>
  invoke<CloudPolicyStatus>("cloud_policy_bind", { request });

export const refreshCloudPolicy = (request: CloudPolicyCredentialRequest) =>
  invoke<void>("cloud_policy_refresh", { request });
export const lockCloudPolicy = () => invoke<void>("cloud_policy_lock");
export const unbindCloudPolicy = (organizationId: string) =>
  invoke<CloudPolicyStatus>("cloud_policy_unbind", { request: { organizationId } });
export const cloudPolicyStatus = (organizationId: string) =>
  invoke<CloudPolicyStatus>("cloud_policy_status", { request: { organizationId } });

File: src-tauri/src/commands/cloud_policy.rs
use tauri::State;

use crate::cloud::{
    CloudPolicyBindingRequest, CloudPolicyCredentialRequest, CloudPolicyOrganizationRequest,
    CloudPolicyService, CloudPolicyStatus,
};
use crate::domain::AppResult;

#[tauri::command]
pub async fn cloud_policy_bind(
    request: CloudPolicyBindingRequest,
    policies: State<'_, CloudPolicyService>,
) -> AppResult<CloudPolicyStatus> {
    policies.bind(request).await
}

#[tauri::command]
pub async fn cloud_policy_refresh(
    request: CloudPolicyCredentialRequest,
    policies: State<'_, CloudPolicyService>,
) -> AppResult<()> {
    policies.refresh(request).await
}

#[tauri::command]
pub async fn cloud_policy_lock(policies: State<'_, CloudPolicyService>) -> AppResult<()> {
    policies.lock().await;
    Ok(())
}

#[tauri::command]
pub async fn cloud_policy_unbind(
    request: CloudPolicyOrganizationRequest,
    policies: State<'_, CloudPolicyService>,
) -> AppResult<CloudPolicyStatus> {
    policies.unbind(request.organization_id).await
}

#[tauri::command]
pub async fn cloud_policy_status(
    request: CloudPolicyOrganizationRequest,
    policies: State<'_, CloudPolicyService>,
) -> AppResult<CloudPolicyStatus> {
    policies.status(request.organization_id).await
}
